File: lib/herb/analysis/ruby_locals_index/reference_collector.rb
# frozen_string_literal: true

require "prism"

require_relative "named_reference"

module Herb
  module Analysis
    class RubyLocalsIndex
      class ReferenceCollector
        attr_reader :local_reads #: Array[NamedReference]
        attr_reader :bare_calls #: Array[NamedReference]
        attr_reader :parameters #: Array[NamedReference]
        attr_reader :assignments #: Array[NamedReference]

        #: (Prism::Node) -> void
        def initialize(program)
          @local_reads = []
          @bare_calls = []
          @parameters = []
          @assignments = []

          walk(program)
        end

        private

        #: (Prism::Node) -> void
        def walk(node)
          record(node)

          node.compact_child_nodes.each { |child| walk(child) }
        end

        #: (Prism::Node) -> void
        def record(node)
          record_read(node)
          record_binding(node)
        end

        #: (Prism::Node) -> void
        def record_read(node)
          case node
          when Prism::LocalVariableReadNode
            add(@local_reads, node.name, node.location)
          when Prism::CallNode
            add(@bare_calls, node.name, node.message_loc) if bare_call?(node)
          end
        end

        #: (Prism::Node) -> void
        def record_binding(node)
          case node
          when Prism::RequiredParameterNode
            add(@parameters, node.name, node.location)
          when Prism::OptionalParameterNode, Prism::BlockParameterNode, Prism::RestParameterNode
            add(@parameters, node.name, node.name_loc)
          when Prism::LocalVariableTargetNode
            add(@assignments, node.name, node.location)
          when Prism::LocalVariableWriteNode, Prism::LocalVariableAndWriteNode,
               Prism::LocalVariableOrWriteNode, Prism::LocalVariableOperatorWriteNode
            add(@assignments, node.name, node.name_loc)
          end
        end

        #: (Prism::CallNode) -> bool
        def bare_call?(node)
          node.receiver.nil? && node.arguments.nil? && node.block.nil? && !node.name.to_s.end_with?("=")
        end

        #: (Array[NamedReference], Symbol?, Prism::Location?) -> void
        def add(references, name, location)
          return unless name && location

          references << NamedReference.new(name.to_s, location.start_offset, location.length)
        end
      end
    end
  end
end

File: rust/herb-analysis/src/ruby_locals_index/reference_collector.rs
use herb::prism::PrismNode;

use super::named_reference::NamedReference;

const LOCAL_VARIABLE_READ: &str = "LocalVariableReadNode";
const CALL_NODE: &str = "CallNode";
const REQUIRED_PARAMETER: &str = "RequiredParameterNode";
const OPTIONAL_PARAMETER: &str = "OptionalParameterNode";
const BLOCK_PARAMETER: &str = "BlockParameterNode";
const REST_PARAMETER: &str = "RestParameterNode";
const LOCAL_VARIABLE_TARGET: &str = "LocalVariableTargetNode";

const LOCAL_WRITE_NODES: [&str; 4] = [
  "LocalVariableWriteNode",
  "LocalVariableAndWriteNode",
  "LocalVariableOrWriteNode",
  "LocalVariableOperatorWriteNode",
];

const SIGIL_PARAMETERS: [&str; 2] = [BLOCK_PARAMETER, REST_PARAMETER];

#[derive(Debug, Default)]
pub struct ReferenceCollector {
  pub local_reads: Vec<NamedReference>,
  pub bare_calls: Vec<NamedReference>,
  pub parameters: Vec<NamedReference>,
  pub assignments: Vec<NamedReference>,
}

impl ReferenceCollector {
  pub fn new(program: &PrismNode) -> Self {
    let mut collector = Self::default();
    collector.walk(program);

    collector
  }

  fn walk(&mut self, node: &PrismNode) {
    self.record(node);

    for child in &node.children {
      self.walk(child);
    }
  }

  fn record(&mut self, node: &PrismNode) {
    self.record_read(node);
    self.record_binding(node);
  }

  fn record_read(&mut self, node: &PrismNode) {
    if node.is(LOCAL_VARIABLE_READ) {
      if let Some(reference) = whole_node(node) {
        self.local_reads.push(reference);
      }
    } else if node.is(CALL_NODE) && bare_call(node) {
      if let Some(reference) = whole_node(node) {
        self.bare_calls.push(reference);
      }
    }
  }

  fn record_binding(&mut self, node: &PrismNode) {
    if node.is(REQUIRED_PARAMETER) || node.is(LOCAL_VARIABLE_TARGET) {
      if let Some(reference) = whole_node(node) {
        self.parameters_or_assignments(node, reference);
      }

      return;
    }

    if node.is(OPTIONAL_PARAMETER) || SIGIL_PARAMETERS.contains(&node.node_type.as_str()) || LOCAL_WRITE_NODES.contains(&node.node_type.as_str()) {
      let Some(name) = node.name.clone() else {
        return;
      };

      let offset = if SIGIL_PARAMETERS.contains(&node.node_type.as_str()) {
        node.start_offset + 1
      } else {
        node.start_offset
      };

      let length = name.len();
      let reference = NamedReference::new(name, offset, length);

      self.parameters_or_assignments(node, reference);
    }
  }

  fn parameters_or_assignments(&mut self, node: &PrismNode, reference: NamedReference) {
    if node.is(LOCAL_VARIABLE_TARGET) || LOCAL_WRITE_NODES.contains(&node.node_type.as_str()) {
      self.assignments.push(reference);
    } else {
      self.parameters.push(reference);
    }
  }
}

fn whole_node(node: &PrismNode) -> Option<NamedReference> {
  let name = node.name.clone()?;
  let length = name.len();

  Some(NamedReference::new(name, node.start_offset, length))
}

fn bare_call(node: &PrismNode) -> bool {
  node.receiver().is_none() && !node.has_block && node.children.is_empty() && !node.name.as_ref().is_some_and(|name| name.ends_with('='))
}
